lib/Transforms/Scalar/DxilLoopUnroll.cpp

```diff
@@ -633,16 +635,21 @@ static bool BreakUpArrayAllocas(bool AllowOOBIndex, IteratorT ItBegin, IteratorT
   return Success;
 }
 
-static void RecursivelyRemoveLoopFromQueue(LPPassManager &LPM, Loop *L) {
+void DxilLoopUnroll::RecursivelyRemoveLoopOnSuccess(LPPassManager &LPM, Loop *L) {
   // Copy the sub loops into a separate list because
   // the original list may change.
   SmallVector<Loop *, 4> SubLoops(L->getSubLoops().begin(), L->getSubLoops().end());
 
   // Must remove all child loops first.
   for (Loop *SubL : SubLoops) {
-    RecursivelyRemoveLoopFromQueue(LPM, SubL);
+    RecursivelyRemoveLoopOnSuccess(LPM, SubL);
   }
 
+  // Remove any loops/subloops that failed because we are about to
+  // delete the loops, so gotta avoid dangling pointers
+  LoopsThatFailed.erase(L);
```

Comment:
What is the mechanism whereby a loop that failed would get retried?

Reply:
If a loop unroll succeeds, all the blocks and its child loops get deleted. But before that happens new sub Loop structures are created for the cloned blocks (which happens in RecursivelyCreateSubLoopForIteration). Those will get retried.